pkg/volume/empty_dir/empty_dir_quota.go

```diff
@@ -0,0 +1,104 @@
+package empty_dir
+
+import (
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/resource"
+	"k8s.io/kubernetes/pkg/types"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+var _ volume.VolumePlugin = &EmptyDirQuotaPlugin{}
+var _ volume.Builder = &emptyDirQuotaBuilder{}
+
+// EmptyDirQuotaPlugin is a simple wrapper for the k8s empty dir plugin builder.
+type EmptyDirQuotaPlugin struct {
+	// wrapped is the actual k8s emptyDir volume plugin we will pass method calls to.
+	Wrapped volume.VolumePlugin
+
+	// The default quota to apply to each node:
+	DefaultQuota resource.Quantity
+
+	// QuotaApplicator is passed to actual volume builders so they can apply
+	// quota for the supported filesystem.
+	QuotaApplicator QuotaApplicator
+}
+
+func (plugin *EmptyDirQuotaPlugin) NewBuilder(spec *volume.Spec, pod *api.Pod, opts volume.VolumeOptions) (volume.Builder, error) {
+	volBuilder, err := plugin.Wrapped.NewBuilder(spec, pod, opts)
+	if err != nil {
+		return volBuilder, err
+	}
+
+	// Because we cannot access several fields on the k8s emptyDir struct, and
+	// we do not wish to modify k8s code for this, we have to grab a reference
+	// to them ourselves.
+	// This logic is the same as k8s.io/kubernetes/pkg/volume/empty_dir:
+	medium := api.StorageMediumDefault
+	if spec.Volume.EmptyDir != nil { // Support a non-specified source as EmptyDir.
+		medium = spec.Volume.EmptyDir.Medium
+	}
+
+	// Wrap the builder object with our own to add quota functionality:
+	wrapperEmptyDir := &emptyDirQuotaBuilder{
+		wrapped:         volBuilder,
+		pod:             pod,
+		medium:          medium,
+		defaultQuota:    plugin.DefaultQuota,
+		quotaApplicator: plugin.QuotaApplicator,
+	}
+	return wrapperEmptyDir, err
+}
+
+func (plugin *EmptyDirQuotaPlugin) Init(host volume.VolumeHost) error {
+	return plugin.Wrapped.Init(host)
+}
+
+func (plugin *EmptyDirQuotaPlugin) Name() string {
+	return plugin.Wrapped.Name()
```

Comment:
I took another quick attempt at this, I think the problem is in the volume cleaner somewhere @markturansky, If I jump around between master and my branch creating pods with empty dir volumes I start seeing issues like:

E0310 15:46:01.836059   22555 volumes.go:242] Could not create volume cleaner for jenkins-1-centos7-volume-1: can't use volume plugins for 0772ba24-e6f8-11e5-931d-989096de63cb/kubernetes.io~empty-dir: no volume plugin matched

It looked like rolling out a rename was more trouble than it would be worth.